govcd/adminorg.go

```diff
@@ -757,27 +757,42 @@ func (adminOrg *AdminOrg) GetVdcByName(vdcname string) (Vdc, error) {
 
 // QueryCatalogList returns a list of catalogs for this organization
 func (adminOrg *AdminOrg) QueryCatalogList() ([]*types.CatalogRecord, error) {
+	return adminOrg.FindCatalogRecords("")
+}
+
+// FindCatalogRecords given a catalog name, retrieves the catalogRecords for a given organization
+func (adminOrg *AdminOrg) FindCatalogRecords(name string) ([]*types.CatalogRecord, error) {
 	util.Logger.Printf("[DEBUG] QueryCatalogList with org name %s", adminOrg.AdminOrg.Name)
-	queryType := types.QtCatalog
+
 	if adminOrg.client.IsSysAdmin {
-		queryType = types.QtAdminCatalog
+		// Set tenant context
+		adminOrg.client.SetCustomHeader(map[string]string{
```

Comment:
There is a risk of setting header value globally (even for a short term).

The `client` is always used as a pointer - that means it is shared throughout everywhere. Now in case a user is doing any other request before the `defer adminOrg.client.RemoveCustomHeader()` gets into effect - his query would be affected by these headers.

In Terraform case - there could be some other resource doing query - and if it happens to happen at exactly the same time when header is set globally, but not yet removed - its query will be impacted. 

It would be a lot safer to look for an option to directly embed these headers into that particular request.

Reply:
yes, this is very right and thanks for pointing this out. I'll figure out the way to set headers just for this query and not for the whole client.

Reply:
@Didainius I've totally changed the approach here.
I've added several methods `*WithHeaders` that allow to do queries sending headers. I also kept the old one for keeping backwards compatibility and avoid breaking the API.

Having done that, now client has a method called `cumulativeQueryWithHeaders` that allows us to do cumulative queries adding headers, the ones that can be used to set tenant context if needed.

Now, these headers are only sent in this query so we avoid the issues you were pointing out in you last comment.

Reply:
Thanks - I like this very much - just have a run for a few tests to be sure we have no impact.